### DISCUSSION TITLE
Question Regarding Custom Open Graph (OG) Tag Values

### DISCUSSION THREAD

**Discussion by shah-iq** [mearashadowfax/ScrewFast#217] [Q&A] [answered]
I’m currently working with the ScrewFast template for our project and have a question regarding the implementation of Open Graph (OG) tags.

I would like to know if there is a recommended approach for passing custom OG tag values directly within the template. Specifically, I’m interested in how to override the default OG properties for specific pages to ensure that each page has the appropriate metadata for social media sharing.

If there’s documentation available on this topic or if you could provide a brief example, I would greatly appreciate it.

**Comment by mearashadowfax** [comment]
Hey @shah-iq!

I see two potential options for implementing custom OG tags:

1. Passing the data manually (similar to how we handle Google Schema on every page):

You could pass custom OG data directly within the component, like this:

```astro
<MainLayout
  title={pageTitle}
  structuredData={{
    "@context": "https://schema.org",
    "@type": "WebPage",
    // more structured data content
  }}
  ogData={{ 
    // your custom OG content here 
  }}
>
```

This would require some updates to the `Meta.astro` component to handle the `ogData` prop and render the appropriate OG tags.

2. Use an Astro integration:

You could explore one of Astro's integrations to dynamically generate OG data and images. This approach might simplify things, as it doesn't require you to manually adjust the codebase:

https://astro.build/integrations?search=images
https://astro.build/integrations/?search=seo

**Reply by shah-iq** [reply]
@mearashadowfax I’m considering using the Astro SEO integration to manage SEO attributes on our pages. I understand that with the integration, we can directly set the title, description, and other properties using the <AstroSeo /> component.

My question is: If we use the Astro SEO integration to define the title and description for our pages, do we still need to set these attributes using the ScrewFast template's method? Or will the Astro SEO integration handle this entirely, allowing us to skip the additional configuration in the ScrewFast way?

**Reply by mearashadowfax** [reply]
> @mearashadowfax I’m considering using the Astro SEO integration to manage SEO attributes on our pages. I understand that with the integration, we can directly set the title, description, and other properties using the component.
> 
> My question is: If we use the Astro SEO integration to define the title and description for our pages, do we still need to set these attributes using the ScrewFast template's method? Or will the Astro SEO integration handle this entirely, allowing us to skip the additional configuration in the ScrewFast way?

@shah-iq, your understanding is correct! Both options essentially serve the same purpose in the same way. In the ScrewFast template, you're manually setting the title and structured data directly inside your layout on each page. Similarly, with the Astro SEO integration, you are also defining these properties, but via the `<SEO>` component.

**Reply by shah-iq** [reply]
> > @mearashadowfax I’m considering using the Astro SEO integration to manage SEO attributes on our pages. I understand that with the integration, we can directly set the title, description, and other properties using the component.
> > My question is: If we use the Astro SEO integration to define the title and description for our pages, do we still need to set these attributes using the ScrewFast template's method? Or will the Astro SEO integration handle this entirely, allowing us to skip the additional configuration in the ScrewFast way?
> 
> @shah-iq, your understanding is correct! Both options essentially serve the same purpose in the same way. In the ScrewFast template, you're manually setting the title and structured data directly inside your layout on each page. Similarly, with the Astro SEO integration, you are also defining these properties, but via the `<SEO>` component.

Thanks for the clarification and the explanation.

**I have another question:** How can we specify meta tags in the markdown files? I noticed there’s an option to set the meta title, but I couldn’t find a way to update the meta description. Is there a specific syntax or method we should follow to include the meta description in the markdown files?

**Reply by mearashadowfax** [reply]
> **I have another question:** How can we specify meta tags in the markdown files? I noticed there’s an option to set the meta title, but I couldn’t find a way to update the meta description. Is there a specific syntax or method we should follow to include the meta description in the markdown files?

You mean for each individual blog post page, at [src/pages/blog/[...slug].astro](https://github.com/mearashadowfax/ScrewFast/blob/2143cc34bcc86f3fd2a39541f6ae2c8c40d1e509/src/pages/blog/%5B...slug%5D.astro), for example? With current implementation, should look like this:
```astro
// src/pages/blog/[...slug].astro
---
const pageTitle: string = `${post.data.title} | ${SITE.title}`;
---

<MainLayout
  title={pageTitle}
  structuredData={{
    "@context": "https://schema.org",
    "@type": "WebPage",
    "@id": "https://screwfast.uk",
    url: `https://screwfast.uk/blog/${post.slug}`,
    name: `${pageTitle}`,
    description: `${post.data.description}`,
    isPartOf: {
      "@type": "WebSite",
      url: "https://screwfast.uk",
      name: "ScrewFast",
      description: "",
    },
    inLanguage: "en",
  }}
>
```

By doing this, the title and description from your markdown frontmatter will automatically populate the meta title and description for each blog post page.

You can also extend the frontmatter section of each markdown file to include a field you would like to add to meta tags: [src/content/blog/en/post-1.md](https://github.com/mearashadowfax/ScrewFast/blob/2143cc34bcc86f3fd2a39541f6ae2c8c40d1e509/src/content/blog/en/post-1.md)

**Reply by shah-iq** [reply]
> You mean for each individual blog post page, at [src/pages/blog/[...slug].astro](https://github.com/mearashadowfax/ScrewFast/blob/2143cc34bcc86f3fd2a39541f6ae2c8c40d1e509/src/pages/blog/%5B...slug%5D.astro), for example?

Yes, you are right. The description was not getting passed to the `MainLayout.astro` file and from that to the `Meta.astro`, so, what I did is to pass the description when calling the `MainLayout` component from the [src/pages/blog/[...slug].astro](https://github.com/mearashadowfax/ScrewFast/blob/2143cc34bcc86f3fd2a39541f6ae2c8c40d1e509/src/pages/blog/%5B...slug%5D.astro) page. Attached a screenshot for your reference

![image](https://github.com/user-attachments/assets/dd12d98f-2808-449c-a125-c294cd00be1a)

**Reply by shah-iq** [reply]
@mearashadowfax For the blog posts (like [src/content/blog/en/post-1.md](https://github.com/mearashadowfax/ScrewFast/blob/2143cc34bcc86f3fd2a39541f6ae2c8c40d1e509/src/content/blog/en/post-1.md)), we want the meta title to be different from the page title for SEO purposes. Instead of using the title in the `<title>` tag like this: `<title>{title}</title>`. 

We’d like to use the meta title value from the frontmatter of the markdown files. How can we pass the meta title from the frontmatter and use it for the meta title in the `MainLayout.astro` file, while still keeping the title for other purposes?

**Reply by mearashadowfax** [reply]
> @mearashadowfax For the blog posts (like [src/content/blog/en/post-1.md](https://github.com/mearashadowfax/ScrewFast/blob/2143cc34bcc86f3fd2a39541f6ae2c8c40d1e509/src/content/blog/en/post-1.md)), we want the meta title to be different from the page title for SEO purposes. Instead of using the title in the `<title>` tag like this: `<title>{title}</title>`.
> 
> We’d like to use the meta title value from the frontmatter of the markdown files. How can we pass the meta title from the frontmatter and use it for the meta title in the `MainLayout.astro` file, while still keeping the title for other purposes?

If I understood you correctly, here's what needs to be done.
1. Add a `metaTitle` field in the frontmatter of the blog post.

```md
---
title: "Maximizing Efficiency with ScrewFast's Cutting-Edge Tools"
metaTitle: "Cutting-Edge Tools by ScrewFast | Innovating Construction Efficiency"
```

2.  Modify `MainLayout.astro` to use `metaTitle`:

```astro
const { title = SITE.title, meta, structuredData, lang = "en", metaTitle } = Astro.props;

interface Props {
  title?: string;
  meta?: string;
  structuredData?: object;
  lang?: string;
  metaTitle?: string;
}
---

<html lang={lang} class="scrollbar-hide lenis lenis-smooth scroll-pt-16">
  <head>
    <Meta meta={meta} structuredData={structuredData} />
    <title>{metaTitle || title}</title>
  </head>
```

3. Update the Blog Post Page to pass `metaTitle`:

```astro
---
import MainLayout from "@/layouts/MainLayout.astro";
import { SITE } from "@data/constants";

const { post } = Astro.props;

const pageTitle = `${post.data.title} | ${SITE.title}`;
const metaTitle = post.data.metaTitle || post.data.title; // Fallback to post title if metaTitle is not defined
---

<MainLayout title={pageTitle} metaTitle={metaTitle}>
  <!-- Blog stuff -->
</MainLayout>
```

It will fallback to `title` if `metaTitle` is not available.

**Reply by shah-iq** [reply]
@mearashadowfax following this approach I encountered an error when using `const metaTitle = post.data.metaTitle || post.data.title;` with the message that `Property 'metaTitle' does not exist on type...`. To resolve this, I added `metaTitle: z.string().optional()`, in the `config.ts` file under the `blogCollection` definition. After that, I was able to access the `metaTitle` value from the markdown files in the Astro components.

Is this the correct way to handle this error, or would you recommend a different approach?

**Reply by mearashadowfax** [reply]
> Is this the correct way to handle this error, or would you recommend a different approach?

This is a solid approach, and you're handling it correctly.

**Comment by mearashadowfax** [comment, marked as answer]
For the first option, it would look like this:

```astro
// Meta.astro
---
// Default properties for the Meta component
const defaultProps = {
  meta: SITE.description,
  structuredData: SEO.structuredData,
  ogData: {
    title: OG.title,
    description: OG.description,
    image: OG.image,
    url: SITE.url,
  },
};

// Extract props and set defaults if not provided
const {
  meta = defaultProps.meta,
  structuredData = defaultProps.structuredData,
  ogData = defaultProps.ogData,
} = Astro.props;

// Define metadata for OG tags
const URL = `${Astro.site}`; // Site URL from astro.config.mjs
const ogTitle = ogData.title;
const ogDescription = ogData.description;
const socialImageRes = await getImage({
  src: ogData.image, // Use custom image if provided
  width: 1200,
  height: 600,
});
const socialImage = Astro.url.origin + socialImageRes.src;

---
<!-- Open Graph tags -->
<meta property="og:locale" content="en_US" />
<meta property="og:url" content={ogData.url || URL} />
<meta property="og:type" content="website" />
<meta property="og:title" content={ogTitle} />
<meta property="og:description" content={ogDescription} />
<meta property="og:image" content={socialImage} />
<meta property="og:image:width" content="1200" />
<meta property="og:image:height" content="600" />
<meta property="og:image:type" content="image/png" />

<!-- Twitter tags -->
<meta name="twitter:card" content="summary_large_image" />
<meta property="twitter:url" content={ogData.url || URL} />
<meta name="twitter:title" content={ogTitle} />
<meta name="twitter:description" content={ogDescription} />
<meta name="twitter:image" content={socialImage} />
```

In the `MainLayout.astro`, we can pass the custom Open Graph data as ogData, similar to how we are passing structuredData.

And finally, on each individual page:
```astro
<MainLayout
  title={pageTitle}
  structuredData={{
    "@context": "https://schema.org",
    // etc
  }}
  ogData={{
    title: "ScrewFast Services – Expert Consultation | ScrewFast",
    description: "Get top-notch consultation services from ScrewFast experts. We help you with all your construction and hardware tool needs.",
    image: "/path to OG image  for this page",
    url: "https://screwfast.uk/services",
  }}
>
```

**Reply by shah-iq** [reply]
Thanks @mearashadowfax.
